### PR TITLE
search: factor out Zoekt global query creation

### PR DESCRIPTION
**Pull request by rvantonder**
Stacked on https://github.com/sourcegraph/sourcegraph/pull/23935.

This breaks up `func zoektSearchGlobal` into the part that creates the global query (which only needs a subset of `ZoektParameter` args, not all of it) and the part that executes the query. This allows a lot of logic simplifications up the stack, keep following.

### CODE REVIEW

File: internal/search/zoekt/indexed_search.go [hunk 1]
```diff
@@ -321,62 +321,58 @@ func NewIndexedSearchRequest(ctx context.Context, args *search.TextParameters, t
...
 // simpler because we don't have to resolve revisions before sending off (global)
 // requests to Zoekt.
-func zoektSearchGlobal(ctx context.Context, args *search.ZoektParameters, c streaming.Sender) error {
-	if args == nil {
-		return nil
-	}
-
-	if args.Mode != search.ZoektGlobalSearch {
-		return fmt.Errorf("zoektSearchGlobal called with args.Mode %d instead of %d", args.Mode, search.ZoektGlobalSearch)
+func zoektGlobalQuery(ctx context.Context, query zoektquery.Q, mode search.GlobalSearchMode, repoOptions search.RepoOptions, userPrivateRepos []types.RepoName) (zoektquery.Q, error) {
```
Comment (rvantonder) [inline internal/search/zoekt/indexed_search.go:333]: function now returns the global `zoektquery.Q`

File: internal/search/zoekt/indexed_search.go [hunk 2]
```diff
@@ -321,62 +321,58 @@ func NewIndexedSearchRequest(ctx context.Context, args *search.TextParameters, t
...
 		head := []string{"HEAD"}
-		for _, r := range args.UserPrivateRepos {
+		for _, r := range userPrivateRepos {
 			privateRepoSet[string(r.Name)] = head
 		}
-		qs = append(qs, zoektquery.NewAnd(&zoektquery.RepoBranches{Set: privateRepoSet}, args.Query))
+		qs = append(qs, zoektquery.NewAnd(&zoektquery.RepoBranches{Set: privateRepoSet}, query))
 	}
 
-	return doZoektSearchGlobal(ctx, zoektquery.Simplify(zoektquery.NewOr(qs...)), args, c)
```
Comment (rvantonder) [inline internal/search/zoekt/indexed_search.go:374]: Key change, we don't run Zoekt here, we just return the query. `doZoektSearchGlobal` requires state in `args` that are quite different from the state needed to construct the Zoekt query.

File: internal/search/zoekt/indexed_search.go [hunk 3]
```diff
@@ -436,7 +432,11 @@ func zoektSearch(ctx context.Context, args *search.ZoektParameters, repos *Index
 	}
 
 	if args.Mode == search.ZoektGlobalSearch {
-		return zoektSearchGlobal(ctx, args, c)
+		q, err := zoektGlobalQuery(ctx, args.Query, args.Mode, args.RepoOptions, args.UserPrivateRepos)
+		if err != nil {
+			return err
+		}
+		return doZoektSearchGlobal(ctx, q, args.Typ, args.Zoekt.Client, args.FileMatchLimit, args.Select, c)
```
Comment (rvantonder) [inline internal/search/zoekt/indexed_search.go:439]: Key change: this is where we call our function that returns the query that's been modified to run a global search. Then we execute it. We can see here, the separation of state required for constructing the query:

```
args.Mode, args.RepoOptions, args.UserPrivateRepos
```

versus the state needed to just run the query:

```
query, args.Typ, args.Zoekt.Client, args.FileMatchLimit, args.Select
```

In fact, they are disjoint!

**Review by stefanhengl — APPROVED** [review]

**Review by camdencheek — APPROVED** [review]

File: internal/search/zoekt/indexed_search.go [hunk 4]
```diff
@@ -321,62 +321,58 @@ func NewIndexedSearchRequest(ctx context.Context, args *search.TextParameters, t
...
 // requests to Zoekt.
-func zoektSearchGlobal(ctx context.Context, args *search.ZoektParameters, c streaming.Sender) error {
-	if args == nil {
-		return nil
-	}
-
-	if args.Mode != search.ZoektGlobalSearch {
-		return fmt.Errorf("zoektSearchGlobal called with args.Mode %d instead of %d", args.Mode, search.ZoektGlobalSearch)
+func zoektGlobalQuery(ctx context.Context, query zoektquery.Q, mode search.GlobalSearchMode, repoOptions search.RepoOptions, userPrivateRepos []types.RepoName) (zoektquery.Q, error) {
+	if mode != search.ZoektGlobalSearch {
```
[the remark this answers is not in the document]
Comment (rvantonder) [reply internal/search/zoekt/indexed_search.go:334]: Hopefully this is answered after reviewing the rest 😄 . I'm going incrementally and it's important to keep changes small so things are tractable. Then it becomes extremely obvious when cruft isn't needed.

**Pull request merged**